Migrate GslCore to dotnet standard + FAKE 5

legezam opened the pull request:
This pull request aims to migrate Amyris.GslCore to dotnet standard and FAKE 5. I used Dmx.GslCore as a base but i did the migration from scratch to make sure everything is okay.

Please note, in order to build this branch, you may need to remove `bin` and `obj` directories and sometimes i even had to clear nuget local caches (`dotnet nuget locals all --clear`)

## Breaking changes

During the migration it turned out that some functionality cannot be migrated directly anymore:
- SourceLink support - as per my understanding, it became a dotnet cli tool and is no longer usable from fake. If it is necessary to have, we need to find a workaround
- Generation of documentation - the new version of FSharp.Formatting that supports dotnetstandard uses a different `generate.fsx` that doesn't rely on fakelib. Also it generates both REFERENCE and HELP at the same time, so it no longer makes sense to have them as separate build tasks. I tried to implement something that mostly resembles the original produced documentation but based on your feedback it may require more work on my side
- The octokit.fsx is not supported in the new fake, however there is a new module, called GitHub which is mostly the same. As the original octokit based release task had the actual file upload commented out, it was unfunctional. Therefore i did a simple 1-1 api conversion in the task and never actually tested.

## Recommended process of Amyris.GslCore update

As discussed with @daz10000 , Amyris may want AmyrisBio library to be migrated in parallel with GslCore. As migration of AmyrisBio is not 100% complete on my side, i did a small change in GslCore: it relies on Dmx.AmyrisBio library temporarily. (it is dotnetstandard) Just to make it compile for now.

I would imagine this update process for the two libraries:
1. whenever you accept this PR, do not release (as it has Dmx.AmyrisBio as a dependency)
2. i submit a PR for upgrading AmyrisBio to dotnetstandard + FAKE5 (*UPDATE:* submitted the [PR](https://github.com/Amyris/AmyrisBio/pull/4))
3. release AmyrisBio
4. i am submitting another PR to GslCore where i change Dmx.AmyrisBio -> AmyrisBio
5. release GslCore

This way of updating ensures that the two updates won't interfere, can be worked on in parallel and all repositories are always possible to compile.

Please note, this is just a proposed solution, i am happy to do in any other preferred way.

legezam committed:
MNT: adding rider folder to gitignore

legezam committed:
MNT: applying new fsproj structure

Side effect: Amyris.Bio -> Dmx.Amyris.Bio (since Amyris.Bio is not yet migrated)

legezam committed:
MNT: migrated project to dotnet core

legezam committed:
MNT: updated FSharp.Formatting

legezam committed:
MNT: updated to FAKE 5

legezam committed:
OOPS: some changes were left out

legezam committed:
MNT: removed sourcelink

legezam committed:
MNT: migrating to new "fake boostrapper"

legezam committed:
MNT: updating FsLexYacc to the very recent version + applied paket simplify

legezam commented on build.fsx:
```diff
@@ -187,95 +188,68 @@ Target "PublishNuget" (fun _ ->
 // Generate the documentation
 
-
-let fakePath = "packages" </> "build" </> "FAKE" </> "tools" </> "FAKE.exe"
```
we no longer need FAKE to invoke FSI as the new FAKE API provides means for invoking FSI directly

legezam commented on src/GslCore/AstAlgorithms.fs:
```diff
@@ -7,7 +7,7 @@ open Amyris.ErrorHandling
 open utils
 
-open Microsoft.FSharp.Text.Lexing
+open FSharp.Text.Lexing
```
in the new FsLexYacc library all namespaces were moved from Microsoft.FSharp.XY to Fsharp.XY, those are the only changes i made

legezam commented on paket.dependencies:
```diff
@@ -1,26 +1,22 @@
 
 nuget FSharp.Core >= 4.1.0 redirects: force
-nuget Amyris.Bio >= 2.0.18 framework: >= net45
+nuget Dmx.Amyris.Bio >= 2.0.23 framework: >= netstandard2.0
```
This has to bechanged back to Amyris.Bio whenever it is migrated

daz10000 commented:
Some notes from experience porting to netcore,  

* you won't need mono anymore for design or runtime.   Instead, just deploy dotnet core runtime from Microsoft or SDK
* Downstream consumers of the library will need to upgrade (e.g. compiler front end, any web services into which GslCore is incorporated)
* Invocation changes - your application ends up being build as a .dll and can be run with dotnet <pathtodll>,  rather an a native .exe.  You wont notice the difference if you're using mono but will be different experience on windows.

legezam committed:
MNT: Moving to dotnet-cli build

- changed MSBuild to dotnet-cli build
- simplified packet.lock (by removing full .net framework packages)
- cleaned packages folder
- including git hash in assemblyinfo
- removing assemblyinfo

chrismacklin commented on build.sh:
```diff
@@ -1,77 +1,27 @@
 #!/usr/bin/env bash
```
I get an error running this script now; previous behavior is to default to building when invoked with no arguments.

```bash
-> DocoptException: errors Error: Error in Argument 1 ('<>')
   Expecting: Command 'build', Command 'run', Flag '--help', Flag '--version' or
   ShortFlag '-h'
```

chrismacklin replied:
Further, when passing the `build` argument to this script:

```
Starting task 'Fsi ': /usr/bin/fsharpi with args --debug:pdbonly --targetprofile:mscorlib docs/tools/generate.fsx --
/home/macklin/src/GslCore> "/usr/bin/fsharpi" --debug:pdbonly --targetprofile:mscorlib docs/tools/generate.fsx -- (In: false, Out: true, Err: true)


/home/macklin/src/GslCore/docs/tools/generate.fsx(2,1): error FS0078: Unable to find the file '../../packages/build/FSharp.Formatting/FSharp.Formatting.fsx' in any of
 /home/macklin/src/GslCore/docs/tools
 /home/macklin/src/GslCore/docs/tools

Finished (Success) 'Fsi ' in 00:00:01.8909871
Finished (Failed) 'GenerateHelp' in 00:00:01.9948433
---------------------------------------------------------------------
Build Time Report
---------------------------------------------------------------------
Target         Duration
------         --------
AssemblyInfo   00:00:01.9929564
GenerateHelp   00:00:01.9941933   (generating reference documentation failed: [])
Build          00:00:00           (skipped)
CopyBinaries   00:00:00           (skipped)
RunTests       00:00:00           (skipped)
GenerateDocs   00:00:00           (skipped)
NuGet          00:00:00           (skipped)
BuildPackage   00:00:00           (skipped)
All            00:00:00           (skipped)
Total:         00:00:04.1852372
Status:        Failure
---------------------------------------------------------------------
Script reported an error:
-> BuildFailedException: Target 'GenerateHelp' failed.
-> One or more errors occurred. (generating reference documentation failed: [])
-> generating reference documentation failed: []
```

legezam replied:
For the first issue: the command line interface of fake has changed starting from v5, please visit: https://fake.build/fake-commandline.html#Basic-examples
basically you should invoke `fake run build.fsx --target x` or `fake build --target x`. If you omit the target, it will run the default target.
As for generate.fsx issue, you have to run paket install before running the build. However, i would recommend getting rid of Fsharp.formatting based documentation completely, because:
- that's the only item that ties us to the existence of a mono
- fragile, cannot be migrated as 1-1
- the generated documentation is not used at all. I mean, it has no much content in it and it is not published anywhere.

Please also note that both fake and paket are now dotnet tools which can be installed with:
```
dotnet tool install --global fake-cli
dotnet tool install --global paket
```

chrismacklin replied:
If that's the case, what is the purpose of `build.sh` in this branch?

chrismacklin replied:
We're not using said documentation, I'd be in favor of removing it from the build.

chrismacklin replied:
In other words: `build.sh` is currently the entrypoint that performs the whole build, by default, with no arguments.  I either need this functionality restored, or a concise recipe for what replaces it.

legezam replied:
> If that's the case, what is the purpose of `build.sh` in this branch?

Basically, it hides the fact that fake is a dotnet cli command from you. It downloads fake for you and passes your provided arguments to it.
The FAKE template comes with it so i thought i retain it for the future as it serves some similar purpose as the old build.sh.
It is here: https://github.com/fsharp/FAKE/blob/e57683b2dfe0fa856502326f308a3a23670ddf46/src/template/fake-template/Content/fake.tool.sh


> We're not using said documentation, I'd be in favor of removing it from the build.

Happy to remove it. See the upcoming commit. Are you okay if i do the same in the Amyris.Bio migration as well?


> In other words: `build.sh` is currently the entrypoint that performs the whole build, by default, with no arguments. I either need this functionality restored, or a concise recipe for what replaces it.

As per my understanding, if you invoke `./build.sh build` or `fake build` without any arguments, it will still run your default target. Is that good for you?

legezam replied:
> > We're not using said documentation, I'd be in favor of removing it from the build.
> 
> Happy to remove it. See the upcoming commit. Are you okay if i do the same in the Amyris.Bio migration as well?

Done. please see my very last commit.

chrismacklin replied:
Onwards to

```
~/src/GslCore$ ./build.sh build
Paket version 5.206.0
The last restore is still up to date. Nothing left to do.
Performance:
 - Runtime: 196 milliseconds
Package Fake.Core.Target is defined more than once in group Main of /home/macklin/src/GslCore/.fake/build.fsx/paket.dependencies
Performance:
 - Cli parsing: 462 milliseconds
 - Packages: 361 milliseconds
 - Script analyzing: 367 milliseconds
 - Runtime: 1 second
There was a problem while setting up the environment:
-> Error during parsing of '/home/macklin/src/GslCore/.fake/build.fsx/paket.lock'.
-> Can't parse ">=". System.ArgumentException: SemVer.Parse
   Parameter name: version
      at Paket.SemVer.Parse$cont@269-1(FSharpList`1 fragments, Unit unitVar)
      at Paket.SemVer.Parse$cont@269(FSharpList`1 fragments, Unit unitVar)
      at Paket.SemVer.f@62-1(String version)
```

legezam replied:
what is your .fake/fake --version ?

chrismacklin replied:
```
~/src/GslCore$ .fake/fake --version
FAKE 5 - F# Make (5.13.5) (this line is written to standard error, see https://github.com/fsharp/FAKE/issues/2066)
FakePath: /home/macklin/src/GslCore/.fake/.store/fake-cli/5.13.5/fake-cli/5.13.5/tools/netcoreapp2.1/any/Fake.Runtime.dll
Paket.Core: 5.203.2
```

legezam replied:
May i ask you to do these steps and try again:
* remove src/GslCore/bin
* remove src/GslCore/obj
* remove tests/GslCore.Tests/bin
* remove tests/GslCore.Tests/obj
*  .paket/paket clear-cache --clear-local
* remove paket-files
* remove .paket
* remove .fake
* remove bin

I know this is basically burning all bridges, but sometimes these tools tend to have something stuck in the pipeline. I am sure it should be something like that because i am trying to do the same in a fresh linux installation with the same tool versions and it works without issues

chrismacklin replied:
I started from a totally clean environment and still ran into the same SemVer issue above.

chrismacklin replied:
(This is on Ubuntu 16.04)

legezam replied:
Unfortunately, i cannot reproduce your issue.
I have pinned the tool + tool package versions in a new commit. Could you repeat the steps above after a pull, please?

Now i tested the branch on Windows 10, Linux Mint 19 and on a vanilla Ubuntu 16.04 using docker.
This builds without any issues:
```docker
FROM ubuntu:16.04

RUN apt-get update && apt-get install -y wget apt-transport-https
RUN wget -q https://packages.microsoft.com/config/ubuntu/16.04/packages-microsoft-prod.deb
RUN dpkg -i packages-microsoft-prod.deb
RUN apt-get update
RUN apt-get install -y apt-transport-https
RUN apt-get install -y dotnet-sdk-2.2 git
RUN git clone https://github.com/legezam/GslCore.git
WORKDIR GslCore
RUN git fetch
RUN git checkout dotnetcore
RUN ./build.sh build
CMD ["bash"]
```

chrismacklin replied:
Thanks for trying to reproduce, I eventually tracked it down to a network-mounted filesystem issue that I'm working to fix now.

legezam replied:
Great! Besides being a repro, this image could serve as a gitlab ci image
for this repo :)

Commit:
MNT: removed FSharp.Formatting , removed docs

daz10000 commented:
Can I put in a plea to at least save the .md files for the docs?  I know they are only a starting point, but I have been documenting any new features and dream one day of catching up.  They are pretty much the only public docs available.  I know Amyris has much better internal docs.  I don't think we need any of the machinery - you can probably just open them in github and browse (gitlab you can at least).  If you want to get rid of them,  I will volunteer to host them somewhere else ;)

legezam commented:
> Can I put in a plea to at least save the .md files for the docs? I know they are only a starting point, but I have been documenting any new features and dream one day of catching up. They are pretty much the only public docs available. I know Amyris has much better internal docs. I don't think we need any of the machinery - you can probably just open them in github and browse (gitlab you can at least). If you want to get rid of them, I will volunteer to host them somewhere else ;)

I was about asking Chris to let us contribute to the wiki section of this repo. (currently it is disabled as i see) That would be a great place for them.

chrismacklin commented:
I'd prefer not to use the github wiki, as it is not easily portable and it would be better for docs to live in the repo proper.

@daz10000 if there are docs in there you want to preserve, we should just keep them in a dedicated `docs` folder and use github to render them.  Do those actually relate to the auto-generated documentation at all?

legezam commented:
@chrismacklin , i am not sure whether you are aware but wiki is just another git repository. :) Those are raw md files in a git that we can move around whenever we want
https://gist.github.com/subfuzion/0d3f19c4f780a7d75ba2

legezam committed:
BUG: adding workaround for bug in fake

https://github.com/fsharp/FAKE/issues/2242

legezam committed:
OOPS: 1 occurrence slipped out the net

daz10000 commented:
> I'd prefer not to use the github wiki, as it is not easily portable and it would be better for docs to live in the repo proper.
> 
> @daz10000 if there are docs in there you want to preserve, we should just keep them in a dedicated `docs` folder and use github to render them. Do those actually relate to the auto-generated documentation at all?

I'd be fine with just putting the useful md files in a top level docs folder for now.  Ok with wiki too but we can take baby steps.  I don't know if the pipeline even renders the md files usefully but they are accessible and rendered by github by default

https://github.com/Amyris/GslCore/blob/master/docs/content/pragmas.md

Commit:
MNT: adding pragma docs back

legezam commented:
> I'd be fine with just putting the useful md files in a top level docs folder for now. Ok with wiki too but we can take baby steps. I don't know if the pipeline even renders the md files usefully but they are accessible and rendered by github by default
> 
> https://github.com/Amyris/GslCore/blob/master/docs/content/pragmas.md

Done :)
https://github.com/Amyris/GslCore/pull/28/commits/2171b66865a571ffc4bd8832ee1e096dac5bed2e#diff-5aff3747cdb55428863717f9b9fb82e9

Commit:
MNT: synchronizing versions with tools

legezam commented:
Great! Besides being a repro, this image could serve as a gitlab ci image
for this repo :)

On Wed, May 15, 2019, 00:13 Chris Macklin <notifications@github.com> wrote:

> *@chrismacklin* commented on this pull request.
> ------------------------------
>
> In build.sh
> <https://github.com/Amyris/GslCore/pull/28#discussion_r284020953>:
>
> > @@ -1,77 +1,27 @@
>  #!/usr/bin/env bash
>
> Thanks for trying to reproduce, I eventually tracked it down to a
> network-mounted filesystem issue that I'm working to fix now.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Amyris/GslCore/pull/28?email_source=notifications&email_token=AA6KJVYZFTHJRBSRJ6ZCGQTPVM2PNA5CNFSM4HGSL45KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBYUGIIA#discussion_r284020953>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6KJV2LOMOWJQGTH47NCQLPVM2PNANCNFSM4HGSL45A>
> .
>

Commit:
MNT: getting rid of FParsec + pinning FSLexYacc as 9.x.x has issues

chrismacklin commented:
What tooling replaces what was previously accomplished by running `./build.sh NuGet`?  Previously this would determine the version number from the release notes and build an appropriate package in the bin folder; PublishNuget would do this and also push to nuget.

legezam commented:
> 
> 
> What tooling replaces what was previously accomplished by running `./build.sh NuGet`? Previously this would determine the version number from the release notes and build an appropriate package in the bin folder; PublishNuget would do this and also push to nuget.

@chrismacklin , it should behave exactly the same as before. I tested `Nuget` myself, however, i couldn't test `publishnuget` as i don't have access.

the new form of `./build.sh NuGet` is now `./build.sh build --target NuGet`

chrismacklin commented:
The --target argument is what was missing, thanks.

Yzupnick reviewed (changes requested):
So far so good. Thanks for this! 
See the AmyrisBio PR regarding order of operations for merging and release.

# Things to do in this PR
- [x] Update `paket.dependencies` back to Amyris.Bio
- [ ] Rebase on master and resolve conflicts

Thanks!

legezam commented:
> 
> 
> So far so good. Thanks for this!
> See the AmyrisBio PR regarding order of operations for merging and release.
> # Things to do in this PR
> 
>     * [ ]  Update `paket.dependencies` back to Amyris.Bio
> 
>     * [ ]  Rebase on master and resolve conflicts
> 
> 
> Thanks!

Hi, Checked your proposal, i am absolutely okay with it. Please let me know once you released the amyris.bio lib and will update this PR accordingly. Thanks!

Commit:
MNT: upgraded to netcore3.1 and netstandard2.1

legezam commented:
@Yzupnick , please note, i did the same here as here: https://github.com/Amyris/AmyrisBio/pull/4#issuecomment-566533412

please let me know if i can replace the dmx amyris lib with the updated amyris lib. 

Thanks!

Commit:
MNT: reverted Dmx.Amyris.Bio package back to the newly released Amyris.Bio package

Yzupnick commented:
Looks like there is still a conflict preventing the merge.

legezam commented:
> Looks like there is still a conflict preventing the merge.

Indeed, there were changes on your master since my last rebase. Did rebase again, should be fine now!
